test/integration/aws_metadata_fetcher_integration_test.cc

```diff
@@ -0,0 +1,152 @@
+  AwsMetadataIntegrationTestSuccess() : AwsMetadataIntegrationTestBase(200, 0) {}
+};
+
+TEST_F(AwsMetadataIntegrationTestSuccess, Success) {
+  const auto endpoint = fmt::format("{}:{}", Network::Test::getLoopbackAddressUrlString(version_),
+                                    lookupPort("listener_0"));
+  const auto response = Utility::metadataFetcher(endpoint, "", "");
+
+  ASSERT_TRUE(response.has_value());
+  EXPECT_EQ("METADATA_VALUE", *response);
+
+  ASSERT_NE(nullptr, test_server_->counter("http.metadata_test.downstream_rq_completed"));
+  EXPECT_EQ(1, test_server_->counter("http.metadata_test.downstream_rq_completed")->value());
+}
+
+TEST_F(AwsMetadataIntegrationTestSuccess, AuthToken) {
+  const auto endpoint = fmt::format("{}:{}", Network::Test::getLoopbackAddressUrlString(version_),
+                                    lookupPort("listener_0"));
+  const auto response = Utility::metadataFetcher(endpoint, "", "AUTH_TOKEN");
+
+  ASSERT_TRUE(response.has_value());
+  EXPECT_EQ("METADATA_VALUE_WITH_AUTH", *response);
+
+  ASSERT_NE(nullptr, test_server_->counter("http.metadata_test.downstream_rq_completed"));
+  EXPECT_EQ(1, test_server_->counter("http.metadata_test.downstream_rq_completed")->value());
+}
+
+class AwsMetadataIntegrationTestFailure : public AwsMetadataIntegrationTestBase {
+public:
+  AwsMetadataIntegrationTestFailure() : AwsMetadataIntegrationTestBase(503, 0) {}
+};
+
+TEST_F(AwsMetadataIntegrationTestFailure, Failure) {
+  const auto endpoint = fmt::format("{}:{}", Network::Test::getLoopbackAddressUrlString(version_),
+                                    lookupPort("listener_0"));
+
+  const auto start_time = timeSystem().monotonicTime();
+  const auto response = Utility::metadataFetcher(endpoint, "", "");
+  const auto end_time = timeSystem().monotonicTime();
+
+  EXPECT_FALSE(response.has_value());
+
+  // Verify correct number of retries
+  ASSERT_NE(nullptr, test_server_->counter("http.metadata_test.downstream_rq_completed"));
+  EXPECT_EQ(4, test_server_->counter("http.metadata_test.downstream_rq_completed")->value());
+
+  // Verify correct sleep time between retries: 4 * 1000 = 4000
+  EXPECT_LE(4000,
+            std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count());
+}
+
+class AwsMetadataIntegrationTestTimeout : public AwsMetadataIntegrationTestBase {
+public:
+  AwsMetadataIntegrationTestTimeout() : AwsMetadataIntegrationTestBase(200, 10) {}
+};
+
+TEST_F(AwsMetadataIntegrationTestTimeout, Timeout) {
+  const auto endpoint = fmt::format("{}:{}", Network::Test::getLoopbackAddressUrlString(version_),
+                                    lookupPort("listener_0"));
+
+  const auto start_time = timeSystem().monotonicTime();
+  const auto response = Utility::metadataFetcher(endpoint, "", "");
+  const auto end_time = timeSystem().monotonicTime();
+
+  EXPECT_FALSE(response.has_value());
+
+  // We do now check http.metadata_test.downstream_rq_completed value here because it's
```

Comment:
It looks like downstream_rq_completed is incremented in Mac even in the case connection is terminated on the Curl side due to the timeout. This doesn't happen on Linux.

Unfortunately I don't have access to a Mac machine to be able to debug this in detail.

@dio any suggestions how to proceed with this? Do we have a process when developers do not have access to a specific platform?

Reply:
Interesting. Unfortunately, we only have this CI. I'll try to take a look at this on my local. However, seems like the test is successful now?

Reply:
Yes, because I removed the following check that was succeeding on Linux but failing on Mac:

`
  ASSERT_NE(nullptr, test_server_->counter("http.metadata_test.downstream_rq_completed"));
  EXPECT_EQ(0, test_server_->counter("http.metadata_test.downstream_rq_completed")->value());
`

For, to me unknown, reason, `downstream_rq_completed` was incremented on Mac for each request Curl aborted due to the timeout. On Linux this counter was not incremented.

Reply:
Let's have a comment regarding this for now.